fdbserver/ClusterController.actor.cpp

```diff
@@ -2167,24 +2236,100 @@ ACTOR Future<Void> updateDatacenterVersionDifference( ClusterControllerData *sel
 	}
 }
 
+ACTOR Future<DataDistributorInterface> startDataDistributor( ClusterControllerData *self ) {
+	state Optional<Key> dcId = self->clusterControllerDcId;
+	while ( !self->clusterControllerProcessId.present() || !self->masterProcessId.present() ) {
+		wait( delay(SERVER_KNOBS->WAIT_FOR_GOOD_RECRUITMENT_DELAY) );
+	}
+
+	while (true) {
+		try {
+			while ( self->db.serverInfo->get().recoveryState < RecoveryState::ACCEPTING_COMMITS ) {
+				wait( self->db.serverInfo->onChange() );
+			}
+
+			std::map<Optional<Standalone<StringRef>>, int> id_used = self->usedIds;
+			state WorkerFitnessInfo data_distributor = self->getWorkerForRoleInDatacenter(dcId, ProcessClass::DataDistributor, ProcessClass::NeverAssign, self->db.config, id_used);
+			state InitializeDataDistributorRequest req;
+			req.reqId = g_random->randomUniqueID();
+			TraceEvent("ClusterController_DataDistributorReqID", req.reqId).detail("Recruit", data_distributor.worker.first.address());
+
+			ErrorOr<DataDistributorInterface> distributor = wait( data_distributor.worker.first.dataDistributor.getReplyUnlessFailedFor(req, 1, 0) );
+			if (distributor.present()) {
+				TraceEvent("ClusterController_DataDistributorReqID", req.reqId).detail("Recruited", data_distributor.worker.first.address());
+				return distributor.get();
+			}
+		}
+		catch (Error& e) {
+			TraceEvent("ClusterController_DataDistributorReqID", req.reqId).error(e);
+			if ( e.code() != error_code_no_more_servers ) {
+				throw;
+			}
+		}
+		wait( delay(SERVER_KNOBS->ATTEMPT_RECRUITMENT_DELAY) );
+	}
+}
+
+ACTOR Future<Void> waitDDRejoinOrStartDD( ClusterControllerData *self, ClusterControllerFullInterface *clusterInterface ) {
+	state Future<DataDistributorInterface> newDistributor = Never();
+
+	// wait for a while to see if existing data distributor will join.
+	loop choose {
+		when ( wait( delay(SERVER_KNOBS->WAIT_FOR_DISTRIBUTOR_JOIN_DELAY) ) ) { break; }
+		when ( wait( self->db.serverInfo->onChange() ) ) {  // Rejoins via worker registration
+			if ( self->db.serverInfo->get().distributor.isValid() ) {
+				TraceEvent("ClusterController_InfoChange", self->id).detail("DataDistributorID", self->db.serverInfo->get().distributor.id());
+				break;
+			}
+		}
+	}
+
+	if ( !self->db.serverInfo->get().distributor.isValid() ) {
+		newDistributor = startDataDistributor( self );
+	}
+
+	// Wait on failures and restart it.
+	loop {
+		state Future<Void> distributorFailed = Never();
+		if ( self->db.serverInfo->get().distributor.isValid() ) {
+			distributorFailed = waitFailureClient( self->db.serverInfo->get().distributor.waitFailure, SERVER_KNOBS->DD_FAILURE_TIME );
+		}
+
+		choose {
```

Comment:
change from a choose when to an if statement that either recruits a new distributor or waits for failure